analyzers/src/SonarAnalyzer.CSharp/Rules/UseAwaitableMethod.cs

```diff
@@ -0,0 +1,195 @@
+                        if (awaitableAlternatives.FirstOrDefault() is { Name: { } alternative })
+                        {
+                            nodeContext.ReportIssue(Rule, invocationExpression, alternative);
+                        }
+                    }, SyntaxKind.InvocationExpression);
+                }
+            });
+        });
+
+    private static WellKnownExtensionMethodContainer BuildWellKnownExtensionMethodContainers(Compilation compilation)
+    {
+        var wellKnownExtensionMethodContainer = new WellKnownExtensionMethodContainer();
+        var queryable = compilation.GetTypeByMetadataName(KnownType.System_Linq_Queryable);
+        var enumerable = compilation.GetTypeByMetadataName(KnownType.System_Linq_Enumerable);
+        if (queryable is not null && enumerable is not null)
+        {
+            if (compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFrameworkCore_EntityFrameworkQueryableExtensions) is { } entityFrameworkQueryableExtensions)
+            {
+                wellKnownExtensionMethodContainer.Add(queryable, entityFrameworkQueryableExtensions);
+                wellKnownExtensionMethodContainer.Add(enumerable, entityFrameworkQueryableExtensions);
+            }
+            if (compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFrameworkCore_RelationalQueryableExtensions) is { } relationalQueryableExtensions)
+            {
+                wellKnownExtensionMethodContainer.Add(queryable, relationalQueryableExtensions);
+                wellKnownExtensionMethodContainer.Add(enumerable, relationalQueryableExtensions);
+            }
+        }
+        if (compilation.GetTypeByMetadataName(KnownType.System_Net_Sockets_Socket) is { } socket
+            && compilation.GetTypeByMetadataName(KnownType.System_Net_Sockets_SocketTaskExtensions) is { } socketTaskExtensions)
+        {
+            wellKnownExtensionMethodContainer.Add(socket, socketTaskExtensions);
+        }
+        return wellKnownExtensionMethodContainer;
+    }
+
+    private static ImmutableArray<ISymbol> FindAwaitableAlternatives(WellKnownExtensionMethodContainer wellKnownExtensionMethodContainer, SyntaxNode codeBlock,
+        InvocationExpressionSyntax invocationExpression, SemanticModel semanticModel, ISymbol containingSymbol, CancellationToken cancel)
+    {
+        var awaitableRoot = GetAwaitableRootOfInvocation(invocationExpression);
+        if (awaitableRoot is { Parent: AwaitExpressionSyntax })
+        {
+            return ImmutableArray<ISymbol>.Empty; // Invocation result is already awaited.
+        }
+        if (invocationExpression.EnclosingScope() is { } scope && !IsAsyncCodeBlock(scope))
+        {
+            return ImmutableArray<ISymbol>.Empty; // Not in an async scope
+        }
+        if (semanticModel.GetSymbolInfo(invocationExpression, cancel).Symbol is IMethodSymbol { MethodKind: not MethodKind.DelegateInvoke } methodSymbol
```

Comment:
Can this be cached with a `ConcurrentDictionary<IMethodSymbol, ImmutableArray<ISymbol>>` to avoid further calls if the method was checked by the analyzer before?

Reply:
As discussed offline, this cache would bring little benefit based on the current performance measures.